src/streams.h

```diff
@@ -0,0 +1,169 @@
+    public:
+
+        // default Stream constructor
+        Stream<T>(T head, std::function<Stream<T>()> gen) {
+            Head = head;
+            Tail = nullptr;
+            Gen  = gen;
+        }
+
+        // Return the first element of the Stream
+        T
+        head();
+
+        // Create the tail Stream by removing the head
+        Stream<T>
+        tail() const;
+
+        // Return a Collection with n elements taken from the Stream
+        Collection<T>
+        take(int n);
+
+        // Return a substream of elements of the Stream that match the
+        // predicate function
+        Stream<T>
+        filter(std::function<bool(T)> func) const;
+
+        // Return the Stream that results from the transformation of each
+        // element in the original Stream
+        template<typename Function>
+        Stream<typename std::result_of<Function(T)>::type>
+        map(Function func) const;
+
+    };
+
+    // Return the first element of the Stream
+    template<typename T>
+    T
+    Stream<T>::head() {
+        return Head;
+    }
+
+    // Create the tail Stream by removing the head
+    template<typename T>
+    Stream<T>
+    Stream<T>::tail() const {
+        if (Tail)
+            return *Tail;
+        return Gen();
+    }
+
+    // Return a Collection with n elements taken from the Stream
+    template<typename T>
+    Collection<T>
+    Stream<T>::take(int n) {
+        std::vector<T> list(n);
+        Stream<T> temp = *this;
+        for (int i = 0; i < n; i++) { 
+            list[i] = temp.head();
+            temp = temp.tail();
+        }
+        return Collection<T>(list);
+    }
+
+    // Return a substream of elements of the Stream that match the
+    // predicate function
+    template<typename T>
+    Stream<T>
+    Stream<T>::filter(std::function<bool(T)> func) const {
+        Stream<T> temp = tail();
+    
+        if (func(Head)) {
+            return Stream<T>(Head, [temp, func]() -> Stream<T> {
+                return temp.filter(func);
+            });
+        }
+        return temp.filter(func);
+    }
+
+    // Return the Stream that results from the transformation of each
+    // element in the original Stream
+    template<typename T>
+    template<typename Function>
+    Stream<typename std::result_of<Function(T)>::type>
+    Stream<T>::map(Function func) const {
+        using return_type = typename std::result_of<Function(T)>::type;
+
+        Stream<T> temp = tail();
+        return Stream<return_type>(func(Head), [temp, func]() -> Stream<return_type> {
+            return temp.map(func);
+        });
+    }
+    
+    // Construct a Stream, starting at n, incrementing by step
+    template<typename T>
+    Stream<T>
+    from(T n, T step) {
+        return Stream<T>(n, [=]() -> Stream<T> {
+            return from(n + step, step);
+        });
+    }
+
+    // Construct a Stream, starting at n, incrementing by 1
+    template<typename T>
+    Stream<T>
+    from(T n) {
```

Comment:
we can skip this implementation if you set `step`'s default value to 1 in the definition.